src/profile-logic/profile-data.js

```diff
@@ -594,46 +595,156 @@ export function filterThreadToRange(
   });
 }
 
+// --------------- CallNodePath and CallNodeIndex manipulations ---------------
+
+// Returns a list of CallNodeIndex from CallNodePaths. This function uses a map
+// to speed up the look-up process.
 export function getCallNodeIndicesFromPaths(
   callNodePaths: CallNodePath[],
   callNodeTable: CallNodeTable
 ): Array<IndexIntoCallNodeTable | null> {
+  // This is a Map<CallNodePathHash, IndexIntoCallNodeTable>. This map speeds up
+  // the look-up process by caching every CallNodePath we handle which avoids
+  // looking up parents again and again.
+  const cache = new Map();
   return callNodePaths.map(path =>
-    getCallNodeIndexFromPath(path, callNodeTable)
+    _getCallNodeIndexFromPathWithCache(path, callNodeTable, cache)
   );
 }
 
-export function getCallNodeIndexFromPath(
+// Returns a CallNodeIndex from a CallNodePath, using and contributing to the
+// cache parameter.
+function _getCallNodeIndexFromPathWithCache(
   callNodePath: CallNodePath,
+  callNodeTable: CallNodeTable,
+  cache: Map<string, IndexIntoCallNodeTable>
+): IndexIntoCallNodeTable | null {
+  const hashFullPath = hashPath(callNodePath);
+  let result = cache.get(hashFullPath);
+  if (result !== undefined) {
+    // The cache already has the result for the full path.
+    return result;
+  }
+
+  // This map is basically the result of:
+  // `map.set(index, hashPath(callNodePath.slice(0, index)))`
+  // This is a cache to speed up contributing data to the `cache` map.
+  const sliceHashes = new Map();
+  sliceHashes.set(callNodePath.length, hashFullPath);
+
+  let i = callNodePath.length;
+  let index;
+  while (--i > 0) {
+    // Looking up each parent for this call node, starting from the deepest node.
+    // If we find a parent this makes it possible to start the look up from this location.
+    const subPath = callNodePath.slice(0, i);
+    const hash = hashPath(subPath);
+    index = cache.get(hash);
+    if (index !== undefined) {
+      // Yay, we already have the result for a parent!
+      break;
+    }
+    // Cache the hashed value because we'll need it later.
+    sliceHashes.set(i, hash);
+  }
+
+  result = _getCallNodeIndexFromPath(
+    callNodePath,
+    callNodeTable,
+    index,
+    i,
+    // This function is called at each "step" of the algorithm, so that we can
+    // cache the result for each node to speed up future lookups.
+    (callNodeIndex, indexInCallNodePath) => {
+      // indexInCallNodePath points to the func index that got resolved. That's
+      // why we need this `+ 1` incrementation to get the slice including this
+      // index.
+      // Reminder: this is the cached result of:
+      // `hashPath(callNodePath.slice(0, indexInCallNodePath + 1))`
+      const hash = sliceHashes.get(indexInCallNodePath + 1);
+      if (hash !== undefined) {
```

Comment:
This is the check I was referring to - I'm pretty sure this index is guaranteed to be present in sliceHashes.

Reply:
yeah, Flow is making this mandatory :( I was slightly relying on the fact that the JIT would see that it's always true and optimize for this